### ISSUE TITLE
basically all based are modified

### ISSUE THREAD

**Issue by michieitel** [nanoporetech/megalodon#22]
Hi!

I was trying to call modifications of a PromethIOn run using this command:

```
megalodon \
--guppy-config dna_r9.4.1_450bps_modbases_dam-dcm-cpg_hac_prom.cfg \
--guppy-server-path=/home/ubuntu/tools/ont-guppy/bin/guppy_basecall_server \
--devices 0 1 \
--processes 40 \
--guppy-params "--num_callers 12 --ipc_threads 16" \
--outputs "mappings" "mods" \
--mod-output-formats "bedmethyl" "modvcf" \
--guppy-timeout 10.0 \
--output-directory megalodon \
--mappings-format bam \
--reference $REF \
$fast5_dir
```

Now when I load the resulting bed files into IGV, together with my reference assembly, basically all bases of my assembly are modified. I guess this is a user mistake so I would be happy to get suggestions on the usage. Please specify any log files needed to help.

Thanks
Michael

**Comment by marcus1487** [comment]
The biological context model `dna_r9.4.1_450bps_modbases_dam-dcm-cpg_hac_prom.cfg` was trained only on human and E. coli contexts. Thus results from other sequences contexts have not been trained or tested for this model. When using the model referenced in this issue it is recommended to set `--mod-motif Z CG 0` for CpG methylation or `--mod-motif Z CCWGG 1 --mod-motif Y GATC 1` for E. coli contexts.

For investigation of 5mC and 6mA in all contexts the recently released [all-context model from the Rerio repository](https://github.com/nanoporetech/rerio) is recommended (community note coming shortly). Note though that a calibration file has not yet been added to megalodon for the all-context model. This will be included in the next megalodon minor release.

That being said, it is still surprising that the result would be all bases modified. One avenue to investigate would be to lower the `--mod-binary-threshold`. Setting this value to `0.5` would force all position to make a call on all reads. I would recommend setting the threshold between `0.55` and `0.6` for initial investigation. This can be completed without rerunning megalodon completely by using the `megalodon/scripts/run_aggregation.py` script. For example: `python megalodon/scripts/run_aggregation.py --megalodon-directory megalodon  --outputs mods --output-suffix thresh_0.6 --mod-binary-threshold 0.6 --processes 40`. This will likely increase the noise in the fraction modified estimates, but might improve your results in this case.

We are also working on training methods to produce more robust modified base calling models and will include these in a future Rerio release.

**Comment by michieitel** [comment]
Thanks Marcus!

I tried using the specified rerio all context model:
```
megalodon \
--guppy-config res_dna_r941_min_modbases-all-context_v001.cfg \
--guppy-server-path=/home/ubuntu/tools/ont-guppy/bin/guppy_basecall_server \
--devices 0 1 \
--processes 40 \
--guppy-params "-d /home/ubuntu/tools/rerio/basecall_models" \
--outputs "mappings" "mods" \
--mod-output-formats "bedmethyl" "modvcf" \
--guppy-timeout 10.0 \
--output-directory megalodon \
--mappings-format bam \
--reference $REF \
$fast5_dir

```


Even when setting all GPU related settings (# callers etc) to the defaults I get an GPU memory allocation error:

> 2020-04-06 05:53:38.713014 [guppy/message] ONT Guppy basecall server software version 3.4.5+fb1fbfb
> config file:        /home/ubuntu/tools/rerio/basecall_models/res_dna_r941_min_modbases-all-context_v001.cfg
> model file:         /home/ubuntu/tools/rerio/basecall_models/res_dna_r941_min_modbases-all-context_v001.jsn
> log path:           Aphrocallistes_20171109_1116_MinION_megalodon/guppy_log
> chunk size:         1000
> chunks per runner:  512
> max queued reads:   2000
> num basecallers:    1
> num socket threads: 2
> max returned events: 50000
> gpu device:         cuda:0 cuda:1
> kernel path:        
> runners per device: 4
> 
> 2020-04-06 05:53:41.478444 [guppy/info] CUDA device 0 (compute 7.0) initialised, memory limit 16945512448B (16612786176B free)
> 
> 2020-04-06 05:53:41.645660 [guppy/info] CUDA device 1 (compute 7.0) initialised, memory limit 16945512448B (16612786176B free)
> 
> 2020-04-06 05:53:41.648159 [guppy/info] CUDA device 0 reached memory limit while trying to allocate 0 bytes, currently at 484/16945512448
> 2020-04-06 05:53:41.649350 [guppy/error] CUDA error at /builds/ofan/ont_core_cpp/ont_core/pair_alignment/align_cuda.cpp:481: CUDA_ERROR_INVALID_VALUE
> 2020-04-06 05:53:41.649438 [guppy/warning] An error occurred initialising the basecaller. Aborting.


What could be the reason here?

Thanks
Michael

**Comment by michieitel** [comment]
btw how doe I specify other modifications, e.g. 6AM?

--mod-motif Y AM 0 ???

Can you please provide a full list of modifications and associated megalodon '--mod-motif' settings?

thanks

**Comment by marcus1487** [comment]
For the CUDA error, the Rerio all-context model is only compatible with guppy >= 3.5.1. My guess is that this incompatibility is causing this crash.

For each modified base model the single letter codes (`Z=5mC` and `Y=6mA`) are the only part required from the model for the `--mod-motif` setting. The second argument to `--mod-motif` is a canonical sequence motif defined using [IUPAC ambiguity codes](https://droog.gs.washington.edu/parc/images/iupac.html). The final value is the relative position within the sequence motif for the modified base.

Megalodon does not currently have a script to extract the modified bases encoded within a model, but I will add this as a feature request for a future release. Note though that this will not include the trained motif as this information is not encoded in the model specifications at this time.

**Comment by michieitel** [comment]
> For the CUDA error, the Rerio all-context model is only compatible with guppy >= 3.5.1. My guess is that this incompatibility is causing this crash.

I can confirm the wrong version. But now for some reason I get another error:

> [14:12:30] Loading guppy basecalling backend.
> Traceback (most recent call last):
>   File "/home/ubuntu/tools/anaconda3/envs/megalodon/bin/megalodon", line 11, in <module>
>     sys.exit(_main())
>   File "/home/ubuntu/tools/anaconda3/envs/megalodon/lib/python3.7/site-packages/megalodon/megalodon.py", line 1215, in _main
>     args, mods_info = parse_mod_args(args, model_info)
>   File "/home/ubuntu/tools/anaconda3/envs/megalodon/lib/python3.7/site-packages/megalodon/megalodon.py", line 775, in parse_mod_args
>     if mh.PR_MOD_NAME in args.outputs else None)
>   File "/home/ubuntu/tools/anaconda3/envs/megalodon/lib/python3.7/site-packages/megalodon/megalodon_helper.py", line 277, in get_mod_calibration_fn
>     'config: {}'.format(guppy_config))
> megalodon.megalodon_helper.MegaError: No default modified base calibration file found for guppy config: res_dna_r941_min_modbases-all-context_v001.cfg
>

**Comment by marcus1487** [comment]
This error is the lack of a calibration file for the Rerio all-context model that I mentioned earlier. You can run with `--disable-mod-calibration`, but this will mean that the per-read results will not accurately measure the empirical probability that a base is modified. Thus the `--mod-binary-threshold` will have to be investigated and may not produce optimal results due to issues with scaling the canonical and modified base distributions. Work is ongoing to make this system more robust, but this would be considered a research task at the moment.

**Comment by marcus1487** [comment]
Fixed in release 2.1.
